snuba/api.py

```diff
@@ -181,36 +245,22 @@ def parse_and_run_query(dataset, validated_body, timer):
     if max_days is not None and (to_date - from_date).days > max_days:
         from_date = to_date - timedelta(days=max_days)
 
-    where_conditions = body.get('conditions', [])
-    where_conditions.extend([
+    # XXX: This should already be enforced by the schema at this point.
+    body.setdefault('conditions', [])
+
+    body['conditions'].extend([
         ('timestamp', '>=', from_date),
         ('timestamp', '<', to_date),
     ])
-    where_conditions.extend(dataset.default_conditions())
+
+    body['conditions'].extend(dataset.default_conditions())
 
     # NOTE: we rely entirely on the schema to make sure that regular snuba
     # queries are required to send a project_id filter. Some other special
     # internal query types do not require a project_id filter.
     project_ids = util.to_list(body['project'])
     if project_ids:
-        where_conditions.append(('project_id', 'IN', project_ids))
-
-    having_conditions = body.get('having', [])
-
-    aggregate_exprs = [
-        util.column_expr(dataset, col, body, alias, agg)
-        for (agg, col, alias) in body['aggregations']
-    ]
-    groupby = util.to_list(body['groupby'])
-    group_exprs = [util.column_expr(dataset, gb, body) for gb in groupby]
-
-    selected_cols = [util.column_expr(dataset, util.tuplify(colname), body)
-                     for colname in body.get('selected_columns', [])]
-
-    select_exprs = group_exprs + aggregate_exprs + selected_cols
-    select_clause = u'SELECT {}'.format(', '.join(select_exprs))
-
-    from_clause = u'FROM {}'.format(table)
+        body['conditions'].append(('project_id', 'IN', project_ids))
```

Comment:
This will eventually move to the events dataset.

Reply:
I suspect you may run into problems if you update the body of the query during split queries since the body is passed down by the split_query wrapper to "parse_and_run_query". 
Same above at line 251. Could you please verify these mutations are harmless. Though I think the processing would be more robust if we treated body as immutable.

Reply:
There's no concerns around mutating the body here since this is already deepcopied on line 233.

This was also already mutating the `conditions` list (specifically, the _copy_ of the original `conditions` list from line 233) since the conditions list is created (as a default) during schema validation: https://github.com/getsentry/snuba/blob/2d55ce81766fbc957822bfcea7b47b2156c4b93d/snuba/schemas.py#L92-L104

The branch to set default or return a default value should never be set, other than potentially during tests.

I agree that we should make this immutable moving forward. I think that's part of a larger change, though. The design of the splitter will have to change, and even things like `alias_expr` mutate the body: https://github.com/getsentry/snuba/blob/2d55ce81766fbc957822bfcea7b47b2156c4b93d/snuba/util.py#L187

Reply:
ok, I did not see line 233. If we are not introducing a mutation here. then it is fine

Reply:
Yeah, I think the fact that it's not obvious is a smell in and of itself. I think there are two problems here:

1. If splitter is the component that is concerned with the mutations to the query, _it_ should be the one doing the copy before it yields control to other parts of the codebase, not relying on them to act in its best interests. The way the code is structured right now makes that really easy to get wrong, and I think your concern around it is totally justified.
2. In general the splitter is implemented in a weird part of the query execution pipeline today. If splitting occurred *after* condition rewriting and things (as noted above), it wouldn't be an issue. The splitter would probably still have to make copies of the original payload, but it'd probably be more obvious why.